### PR TITLE
Solid face renderer and related stuff.

### PR DESCRIPTION
**Pull request by lovexyn0827**
#### Added

- Carpet setting solidFaceRenderer
- Carpet setting solidFaceShapePredicate

### CODE REVIEW

**Comment by zly2006** [comment]
I think it is better to use a malilib config instead of carpet's.

**Comment by wafarm** [comment]
MiniHUD has a feature `debugSolidFaceEnabled`, which highlight solid face with red color (screenshot below). What's the difference?
image

**Comment by lovexyn0827** [comment]
> MiniHUD has a feature `debugSolidFaceEnabled`, which highlight solid face with red color (screenshot below). What's the difference?
> image
> 

MiniHUD uses the hidden BlockOutlineRenderer of vanilla client, which renderers all solid faces of all blocks within a 13 by 13 by 13 cube.

But the renderer here only renderers for the face targeted by the player. Besides, the customization of block face shape predication is supported (basically, it means non-full solid faces can be rendered.).

**Review by zly2006 — CHANGES_REQUESTED** [review]
consider using `ConfigOptionList` instead of just a string for predicate mode.

**Comment by lovexyn0827** [comment]
> consider using `ConfigOptionList` instead of just a string for predicate mode.

Should i introduce class `RedenOptionList` as it does not exist and other options use `RedenConfigXXX` instead of classes in vanilla malilib.

**Comment by zly2006** [comment]
> Should i introduce class `RedenOptionList` as it does not exist and other options use `RedenConfigXXX` instead of classes in vanilla malilib.

Yes, that would be good.

**Comment by wafarm** [comment]
> But the renderer here only renderers for the face targeted by the player. Besides, the customization of block face shape predication is supported (basically, it means non-full solid faces can be rendered.).

Got that. btw maybe you should place the code of config options according to their place in the UI, like below line 50.

File: src/main/java/com/github/zly2006/reden/malilib/MalilibSettings.kt [hunk 1]
```diff
@@ -68,5 +68,7 @@ private fun <T : IConfigBase?> ConfigBase<T>.debug() = this.apply(DEBUG_TAB::add
 @JvmField val OPEN_GITHUB_AUTH_SCREEN = RedenConfigHotkey("openGithubAuthScreen", "R,G").debug().hotkey()
 @JvmField val GITHUB_TOKEN = RedenConfigString("githubToken", "").debug()
 @JvmField val ALLOW_SOCIAL_FOLLOW = RedenConfigBoolean("allowSocialFollow", true).debug()
+@JvmField val SOLID_FACE_RENDERER = RedenConfigBoolean("solidFaceRenderer", false).generic()
+@JvmField val SOLID_FACE_SHAPE_PREDICATE = RedenConfigString("solidFaceShapePredicate", "full").generic()
```
Comment (wafarm) [inline src/main/java/com/github/zly2006/reden/malilib/MalilibSettings.kt]: I mean, move them up. (GitHub require comments on what need to be changed in order to suggest changes)

**Review by wafarm — CHANGES_REQUESTED** [review]

**Comment by Qubik65536** [comment]
Follow the PR template next time, otherwise I'll close it, IT IS THERE FOR A REASON.

**Review by wafarm — CHANGES_REQUESTED** [review]
Please remove unused import statements

**Comment by lovexyn0827** [comment]
> Please remove unused import statements

Done

**Review by wafarm — APPROVED** [review]

**Review by zly2006 — APPROVED** [review]

**Review by Qubik65536 — APPROVED** [review]

**Pull request merged**